SPP1-168: Set CBF target to middle (ra, dec) of each scan in OTF mapping

tamerakassie opened the pull request:
Created the  ra dec target to be set on CBF for delay tracking in reverse scan observations, we will create a default target dictionary where the `cbf` key will correspond to the ra dec katpoint target that is fixed and the scans of the antennas will run across at `time_radec` i.e ~ midpoint of a scan line, and the `ants` key will correspond to the `scan_azel_with_nd_trigger`. The interface to `CaptureSession.set_target`is refactored to accept a dictionary of targets instead of a single target, in `reversescan` we create a simple target dictionary which contain the following keys: `cbf`, `ants` with respective  `scan_azel_with_nd_trigger`for the area extent scan and `radec target` for delay tracking on CBF, these will be passed to `CaptureSession`.

To maintain the code flexibility and logic  we use an `instance` checks for the target dictionary to access the `scan_target` or `radec target`  when required by the reverse scan simulations, while not affecting the other scan simulations in  the other `astrokat` tests.

tamerakassie committed:
SPP1-168: Allow different targets to be set

Created the  ra dec target to be set on CBF for delay tracking.
The interface to `CaptureSession.set_target` will accept a dictionary of targets instead of a single target,
in `reversescan` we create a simple target dictionary which contain the following keys: `cbf`, `ants` with
respectic `scan_nd_az_el` for the area extent scan and `radec target`
for delay tracking on CBF.

To maintain the code flexibility we use an `instance` checks for the
target dictionary to access the `scan_target` or `radec target`  when required by the reverse scan simulations, while not affecting the other scans in `astrokat`.

tamerakassie committed:
remove slew-only

ludwigschwardt committed:
SPP1-168 Set target per scan and not per patch

Move the target setting into the scan loop. Use the time roughly in the
middle of the scan (half the scan duration after the start of the current
scan) to determine the (ra, dec) coordinates for the CBF target.

ludwigschwardt committed:
SPP1-168 Flake8 fixes

Instead of documenting the units in a comment, document it in the user
script logs for a change, where it was somehow absent :-)

Write units out in full to match other instances in the session logs.

lanceWilliams commented on astrokat/scans.py:
```diff
@@ -110,14 +110,23 @@ def scan(session, target, nd_period=None, lead_time=None, **kwargs):
     """
+    if isinstance(target, dict):
+        scan_target = target.get('ants')
+        radec_target = target.get('cbf')
```
Could we just call this the 'cbf_target' as opposed to the 'radec_target'? As far as I am aware the 'scan_target' is almost always a 'radec' target anyway.

lanceWilliams commented on astrokat/scans.py:
```diff
@@ -331,8 +339,19 @@ def reversescan(session, target, nd_period=None, lead_time=None, **kwargs):
                          (scanargs["start"][0], scanargs["end"][0]))
+
+        # Determine the cbf targets
+        # use scan_target to the astrometric (ra, dec)
```
This line and 345 read strangely. Are these typos?

ludwigschwardt replied:
It is a bit unclear, yes. I would just say something like:
```
# CBF target is (ra, dec) position of reference antenna in middle of next scan
```

lanceWilliams commented on astrokat/simulate.py:
```diff
@@ -365,21 +369,29 @@ def _target_azel(self, target):
         """
-        az, el = target.azel(simobserver.date)
+        if isinstance(target, dict):
+            scan_target = target.get('ants')
```
I think this simulation function is shared between tracks and scans? Referencing 'scan_targets' here is perhaps misleading.

lanceWilliams commented on astrokat/simulate.py:
```diff
@@ -365,21 +369,29 @@ def _target_azel(self, target):
 
     def _fake_slew_(self, target):
+        if isinstance(target, dict):
+            scan_target = target.get('ants')
```
As the with the comment above, I think this simulation function is shared between tracks and scans? Referencing 'scan_targets' here is perhaps misleading.

lanceWilliams commented on astrokat/observe_main.py:
```diff
@@ -98,7 +98,7 @@ def observe(session, ref_antenna, target_info, **kwargs):
     # simple way to get telescope to slew to target
     if "slewonly" in kwargs:
-        return session.track(target, duration=0.0, announce=False, slew_only=True)
+        return session.track(target, duration=0.0, announce=False)
```
With the proposed changes in katcorelib it seems like the 'slewonly' kwarg wouldn't do anything unless it altered the structure of the target mapping being passed to 'track'?

ludwigschwardt replied:
It does exactly nothing, which is why we need to decide what to do about the discrepancy / interplay between #129, #114 and #113...

lanceWilliams replied:
@ludwigschwardt I think that the changes in [391](https://github.com/ska-sa/katcorelib/pull/391) can reasonably replace what was done in [113](https://github.com/ska-sa/astrokat/pull/113). This "slewonly" check would only need to ensure that it specifically passes track() a target dict with only an 'ants' key. Then there shouldn't be any impact on the rest of the observation setup or execution as far as I can tell.

ludwigschwardt replied:
Should #129 then also effectively enable #114 on site, which hasn't been merged up to now? The `slewonly` functionality has never been used, in effect...

lanceWilliams replied:
Yes, I think so. From distant memory I think this feature came out of trying to improve the startup for the PTUSE observation scripts in particular. I don't think other observations really experienced the same efficiency issue so there was no real pressure for it.

ludwigschwardt committed:
SPP1-168 Fix reversescan unit test

Whoops, the test checks the script logs...

lanceWilliams commented:
I think I am ok with approving https://github.com/ska-sa/katcorelib/pull/391, is this astrokat PR still being worked on/tested?

ludwigschwardt commented:
> I think I am ok with approving [ska-sa/katcorelib#391](https://github.com/ska-sa/katcorelib/pull/391), is this astrokat PR still being worked on/tested?

I think @tamerakassie could still address the minor suggestions you made.

tamerakassie committed:
Addressing PR comments

In this commit I have changed referring of `radec_target` and `scan_target` in the simulate scripts and `reverse_scan` function to the standard keys we set as
`cbf_target` or `ant_target`

ludwigschwardt committed:
SPP1-168 Add note about removed "slew_only" param

ludwigschwardt reviewed (approved):
Happy with this. @lanceWilliams, please just check that Tamera addressed your requested changes.

lanceWilliams reviewed (approved):
Looks ok
